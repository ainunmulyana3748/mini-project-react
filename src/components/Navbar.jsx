import React, { useState } from "react";
import { Link } from "react-router-dom";
import { Menu, X } from "lucide-react";
import { AnimatePresence, motion } from "framer-motion";

const Navbar = () => {
  const token = localStorage.getItem("token");
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);

  const handleLogout = () => {
    localStorage.removeItem("token");
  };

  const toggleMenu = () => {
    setIsMobileMenuOpen((prev) => !prev);
  };

  return (
    <header className="bg-black text-white w-full shadow-md z-50 fixed">
      <div className="max-w-screen-xl mx-auto px-4 py-4 flex items-center justify-between">
        <Link to="/" className="text-xl font-bold text-white">
          Ditemenin.
        </Link>

        {/* Desktop Menu */}
        <nav className="hidden md:flex gap-6 text-sm sm:font-semibold">
          <Link to="/landing-page" className="hover:text-gray-300">
            Home
          </Link>
          <Link to="/" className="hover:text-gray-300">
            Our Tour Guide
          </Link>
        </nav>

        {/* Desktop Auth */}
        <div className="hidden md:block">
          {!token ? (
            <Link
              to="/login"
              className="bg-green-500 hover:bg-green-600 text-white font-medium px-4 py-2 rounded transition text-sm sm:text-base"
            >
              Login
            </Link>
          ) : (
            <Link
              to="/login"
              onClick={handleLogout}
              className="bg-red-500 hover:bg-red-600 text-white font-medium px-4 py-2 rounded transition text-sm sm:text-base"
            >
              Logout
            </Link>
          )}
        </div>

        {/* Mobile Toggle */}
        <button
          className="md:hidden text-white focus:outline-none z-50"
          onClick={toggleMenu}
        >
          {isMobileMenuOpen ? <X size={28} /> : <Menu size={28} />}
        </button>
      </div>

      {/* Mobile Menu Overlay */}
      <AnimatePresence>
        {isMobileMenuOpen && (
          <motion.div
            className="absolute top-full left-0 w-full px-4 py-4 bg-black z-40 space-y-4 text-sm font-medium shadow-lg"
            initial={{ opacity: 0, y: -10, scaleY: 0.95 }}
            animate={{ opacity: 1, y: 0, scaleY: 1 }}
            exit={{ opacity: 0, y: -10, scaleY: 0.95 }}
            transition={{ duration: 0.25, ease: "easeInOut" }}
          >
            <Link
              to="/landing-page"
              className="block text-white hover:text-gray-300"
              onClick={() => setIsMobileMenuOpen(false)}
            >
              Home
            </Link>
            <Link
              to="/"
              className="block text-white hover:text-gray-300"
              onClick={() => setIsMobileMenuOpen(false)}
            >
              Our Tour Guide
            </Link>
            {!token ? (
              <Link
                to="/login"
                className="block bg-green-500 hover:bg-green-600 text-white text-center px-4 py-2 rounded"
                onClick={() => setIsMobileMenuOpen(false)}
              >
                Login
              </Link>
            ) : (
              <Link
                to="/login"
                className="block bg-red-500 hover:bg-red-600 text-white text-center px-4 py-2 rounded"
                onClick={() => {
                  handleLogout();
                  setIsMobileMenuOpen(false);
                }}
              >
                Logout
              </Link>
            )}
          </motion.div>
        )}
      </AnimatePresence>
    </header>
  );
};

export default Navbar;
