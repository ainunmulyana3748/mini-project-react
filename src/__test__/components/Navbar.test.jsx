import { render, screen, fireEvent } from "@testing-library/react";
import Navbar from "../../components/Navbar";
import { BrowserRouter } from "react-router-dom";

// Mock localStorage
const localStorageMock = (() => {
  let store = {};
  return {
    getItem: (key) => store[key] || null,
    setItem: (key, value) => (store[key] = value + ""),
    removeItem: (key) => delete store[key],
    clear: () => (store = {}),
  };
})();
Object.defineProperty(window, "localStorage", { value: localStorageMock });

describe("Navbar Component", () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  test("render brand dan menu desktop (belum login)", () => {
    render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    );
    expect(screen.getByText("Ditemenin.")).toBeInTheDocument();
    expect(screen.getByText("Home")).toBeInTheDocument();
    expect(screen.getByText("Our Tour Guide")).toBeInTheDocument();
    expect(screen.getByText("Login")).toBeInTheDocument();
  });

  test("render tombol Logout jika sudah login", () => {
    window.localStorage.setItem("token", "abc");
    render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    );
    expect(screen.getByText("Logout")).toBeInTheDocument();
  });

  test("klik tombol menu mobile membuka menu", () => {
    render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    );
    const menuBtn = screen.getByRole("button");
    fireEvent.click(menuBtn);
    expect(screen.getAllByText("Home").length).toBeGreaterThan(1); // Ada di mobile menu
  });

  test("klik tombol Logout menghapus token", () => {
    window.localStorage.setItem("token", "abc");
    render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByText("Logout"));
    expect(window.localStorage.getItem("token")).toBeNull();
  });

  test("snapshot navbar", () => {
    const { container } = render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
