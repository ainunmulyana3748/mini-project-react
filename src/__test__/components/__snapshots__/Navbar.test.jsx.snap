// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component snapshot navbar 1`] = `
<div>
  <header
    class="bg-black text-white w-full shadow-md z-50 fixed"
  >
    <div
      class="max-w-screen-xl mx-auto px-4 py-4 flex items-center justify-between"
    >
      <a
        class="text-xl font-bold text-white"
        data-discover="true"
        href="/"
      >
        Ditemenin.
      </a>
      <nav
        class="hidden md:flex gap-6 text-sm sm:font-semibold"
      >
        <a
          class="hover:text-gray-300"
          data-discover="true"
          href="/landing-page"
        >
          Home
        </a>
        <a
          class="hover:text-gray-300"
          data-discover="true"
          href="/"
        >
          Our Tour Guide
        </a>
      </nav>
      <div
        class="hidden md:block"
      >
        <a
          class="bg-green-500 hover:bg-green-600 text-white font-medium px-4 py-2 rounded transition text-sm sm:text-base"
          data-discover="true"
          href="/login"
        >
          Login
        </a>
      </div>
      <button
        class="md:hidden text-white focus:outline-none z-50"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-menu"
          fill="none"
          height="28"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 12h16"
          />
          <path
            d="M4 18h16"
          />
          <path
            d="M4 6h16"
          />
        </svg>
      </button>
    </div>
  </header>
</div>
`;
